src/iop/basecurve.c

```diff
@@ -302,18 +342,340 @@ int process_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, cl_m
+    float *const fine,        // upsampled, blurry output
+    const int wd,             // fine res
+    const int ht)
+{
+  const int cw = (wd-1)/2+1;
+  // fill numbers in even pixels, zero odd ones
+  memset(fine, 0, 4*wd*ht*sizeof(float));
+#ifdef _OPENMP
+#pragma omp parallel for default(none) schedule(static) collapse(2)
+#endif
+  for(int j=0;j<ht;j+=2)
+    for(int i=0;i<wd;i+=2)
+      for(int c=0;c<4;c++)
+        fine[4*(j*wd+i)+c] = 4.0f * input[4*(j/2*cw + i/2)+c];
+
+  // convolve with same kernel weights mul by 4:
+  gauss_blur(fine, fine, wd, ht);
+}
+
+// XXX FIXME: we'll need to pad up the image to get a good boundary condition!
+// XXX FIXME: downsampling will not result in an energy conserving pattern (every 4 pixels one sample)
+// XXX FIXME: neither will a mirror boundary condition (mirrors in subsampled values at random density)
+static inline void gauss_reduce(
+    const float *const input, // fine input buffer
+    float *const coarse,      // coarse scale, blurred input buf
+    float *const detail,      // detail/laplacian, fine scale, or 0
+    const int wd,
+    const int ht)
+{
+  // blur, store only coarse res
+  const int cw = (wd-1)/2+1, ch = (ht-1)/2+1;
+
+  // TODO: pass in output buffer as tmp?
+  float *blurred = dt_alloc_align(64, wd*ht*4*sizeof(float));
+  gauss_blur(input, blurred, wd, ht);
+  for(int j=0;j<ch;j++) for(int i=0;i<cw;i++)
+    for(int c=0;c<4;c++) coarse[4*(j*cw+i)+c] = blurred[4*(2*j*wd+2*i)+c];
+  free(blurred);
+
+#if 0
+  const int cw = (wd-1)/2+1, ch = (ht-1)/2+1;
+  const float w[5] = {1./16., 4./16., 6./16., 4./16., 1./16.};
+  // TODO: border handling!
+  for(int j=1;2*j+2<ht;j++) for(int i=1;2*i+2<wd;i++)
+  { // keep even pixels only
+    const int ind = 4*(j*cw+i), indi = 4*(2*j*wd+2*i);
+    for(int c=0;c<4;c++) coarse[ind + c] = 0.0f;
+    for(int ii=-2;ii<=2;ii++)
+      for(int jj=-2;jj<=2;jj++)
+        for(int c=0;c<4;c++)
+          coarse[ind + c] +=
+            input[indi+4*(wd*jj+ii)+c]*w[jj+2]*w[ii+2];
+  }
+#endif
+  if(detail)
+  {
+    // compute laplacian/details: expand coarse buffer into detail
+    // buffer subtract expanded buffer from input in place
+    gauss_expand(coarse, detail, wd, ht);
+    for(int64_t k=0;k<wd*ht*4;k++)
+      detail[k] = input[k] - detail[k];
+  }
+}
+
 void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ivoid,
-             void *const ovoid, const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+    void *const ovoid, const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
 {
+  const float *const in  = (float *)ivoid;
+  float *const out = (float *)ovoid;
   const int ch = piece->colors;
-  const dt_iop_basecurve_data_t *const d = (const dt_iop_basecurve_data_t *const)(piece->data);
+  dt_iop_basecurve_data_t *const d = (dt_iop_basecurve_data_t *)(piece->data);
+
+  // are we doing exposure fusion?
+  if(d->exposure_fusion)
```

Comment:
Based on recent changes in multiple `process()` variants, i believe turning this `if()` into a function and then conditionally calling it from `process()` would look cleaner.